[Merged by Bors] - feat(ring_theory/ideal): ideal norm evaluation lemmas

Vierkantor opened the pull request:
This PR continues on #17203 by adding a couple lemmas on the ideal norm of some specific ideals:

 * the norm of `⊥` and `⊤`
 * the norm of an ideal `I : ideal S` given an additive isomorphism between `I` and `S`
 * the norm of the ideal generated by `a`
 * the norm of the ideal generated by `insert a s`

It also adds a selection of dependent lemmas. I couldn't find a really neat place for some of them: I think the current places are the least worst but I am very much open to suggestions.

Co-authored-by: Alex J Best <alex.j.best@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #17203

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

riccardobrasca commented on src/linear_algebra/free_module/determinant.lean:
```diff
@@ -0,0 +1,27 @@
+   nontrivial module.
+-/
+
+@[simp] lemma linear_map.det_zero'' {R M : Type*} [comm_ring R] [add_comm_group M] [module R M]
```
This works also for the trivial ring, since it is of the form `... = 0`.

Vierkantor committed:
feat(ring_theory/ideal): ideal norm evaluation lemmas

This PR continues on #17203 by adding a couple lemmas on the ideal norm of some specific ideals:

 * the norm of `⊥` and `⊤`
 * the norm of an ideal `I : ideal S` given an additive isomorphism between `I` and `S`
 * the norm of the ideal generated by `a`
 * the norm of the ideal generated by `insert a s`

It also adds a selection of dependent lemmas. I couldn't find a really neat place for some of them: I think the current places are the least worst but I am very much open to suggestions.

Co-Authored-By: Alex J Best <alex.j.best@gmail.com>

Vierkantor committed:
Don't forget to commit the other files...

riccardobrasca commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -240,3 +243,169 @@ noncomputable def ideal.abs_norm [infinite S] [is_dedekind_domain S]
+@[simp] lemma abs_norm_eq_one_iff {I : ideal S} : abs_norm I = 1 ↔ I = ⊤ :=
+by rw [abs_norm_apply, card_quot_eq_one_iff]
+
+/-- Let `e : S ≃ I` be an additive homomorphism (therefore a `ℤ`-linear equiv).
```
```suggestion
/-- Let `e : S ≃ I` be an additive isomorphism (therefore a `ℤ`-linear equiv).
```

riccardobrasca commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -240,3 +243,169 @@ noncomputable def ideal.abs_norm [infinite S] [is_dedekind_domain S]
+by rw [abs_norm_apply, card_quot_eq_one_iff]
+
+/-- Let `e : S ≃ I` be an additive homomorphism (therefore a `ℤ`-linear equiv).
+Then an alternative way to compute the norm of `I` is given by taking the determinant of `e`. -/
```
```suggestion
Then an alternative way to compute the norm of `I` is given by taking the determinant of `e`. See `nat_abs_det_basis_change` for a more familiar formulation of this result. -/
```
This is maybe subjective, but I think that number theorists know that the norm is the determinant of the base change matrix, but hardly think in term of determinant of linear isomorphisms here.

Vierkantor replied:
```suggestion
Then an alternative way to compute the norm of `I` is given by taking the determinant of `e`.
See `nat_abs_det_basis_change` for a more familiar formulation of this result. -/
```

riccardobrasca commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -240,3 +243,169 @@ noncomputable def ideal.abs_norm [infinite S] [is_dedekind_domain S]
+    ⟨int.nat_abs_ne_zero_of_ne_zero (ideal.smith_coeffs_ne_zero b I hI i)⟩,
+  simp_rw [fintype.card_eq.mpr ⟨(ideal.quotient_equiv_pi_zmod I b hI).to_equiv⟩, fintype.card_pi,
+           zmod.card] ,
+end
```
This is a long proof, but I assume you've already tried to split in smaller pieces...

Vierkantor replied:
Yeah, the alternative would be probably to make `f` a separate definition but I would not expect that that would be useful elsewhere.

Vierkantor replied:
(I pressed the wrong button, now my reply should be there! :P)

riccardobrasca commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -240,3 +243,169 @@ noncomputable def ideal.abs_norm [infinite S] [is_dedekind_domain S]
+end
+
+/-- A basis on `S` gives a basis on `ideal.span {x}`, by multiplying everything by `x`. -/
+noncomputable def basis_span_singleton {ι R S : Type*} [comm_semiring R] [comm_ring S]
```
Can you move this (and the lemmas below) to some other file? I don't know what the most basic file knowing `ideal` and `basis` is thought.

riccardobrasca reviewed (approved):
Can you add that if the norm is prime than the ideal is prime and the corollary about elements/principal ideals? It should be easy. Thanks!

Vierkantor committed:
Clarify docstrings

Co-authored-by: Riccardo Brasca <riccardo.brasca@gmail.com>

Vierkantor committed:
Norm is irreducible iff the ideal is prime

Vierkantor committed:
Move definition lower down the import graph

riccardobrasca commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -240,3 +243,160 @@ noncomputable def ideal.abs_norm [infinite S] [is_dedekind_domain S]
+end
+
+@[simp]
+lemma abs_norm_span_singleton (r : S) :
```
Can you add this theorem, and the multiplicativity of the norm, to a "main results" section in the doc? Thanks!

riccardobrasca reviewed (commented):
Thanks!

bors d+

Vierkantor commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -5,12 +5,15 @@ Authors: Anne Baanen, Alex J. Best
-import linear_algebra.free_module.finite.basic
 import linear_algebra.free_module.pid
+import linear_algebra.free_module.determinant
 import linear_algebra.free_module.ideal_quotient
```
```suggestion
import linear_algebra.free_module.determinant
import linear_algebra.free_module.ideal_quotient
import linear_algebra.free_module.pid
```

Vierkantor commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -5,12 +5,15 @@ Authors: Anne Baanen, Alex J. Best
+import linear_algebra.free_module.determinant
 import linear_algebra.free_module.ideal_quotient
 import linear_algebra.isomorphisms
+import ring_theory.norm
```
```suggestion
```

Vierkantor committed:
Sort / dedup imports

Vierkantor committed:
Add main results to module doc

riccardobrasca commented on src/ring_theory/ideal/norm.lean:
```diff
@@ -240,3 +251,160 @@ noncomputable def ideal.abs_norm [infinite S] [is_dedekind_domain S]
+
+@[simp]
+lemma abs_norm_span_singleton (r : S) :
+  abs_norm (span ({r} : set S)) = (algebra.norm ℤ r).nat_abs :=
```
I wanted to use this result and I have the impression is "morally" wrong, meaning it should use [ring_of_integers.norm](https://leanprover-community.github.io/mathlib_docs/number_theory/number_field/norm.html#ring_of_integers.norm) instead of `algebra.norm ℤ`. The two are of course equal via [rat.ring_of_integers_equiv](https://leanprover-community.github.io/mathlib_docs/number_theory/number_field/basic.html#rat.ring_of_integers_equiv) and I understand that having values in `number_field.ring_of_integers ℚ` rather than `ℤ` is unpleasant, but here is my argument.

The point is that if we replace ℤ with `R` (the ring of integer of a finite extension of `ℚ` and `S` is now the ring of integers of a finite extension `L` of `ℚ` then it's possible that `S` is not free as `R` module, so `norm R : S →* R` can be `0`. I indeed think that the norm relative to `S` should never be used.

It's possible that this would allow to remove the `module.free` assumption in [ideal.is_prime_of_irreducible_abs_norm](https://leanprover-community.github.io/mathlib_docs/ring_theory/ideal/norm.html#ideal.is_prime_of_irreducible_abs_norm) (this is always true if `R = ℤ` but morally it shouldn't be there).

What do you think?

Vierkantor replied:
I think your suggestion is a good idea, as long as we generalize `ring_of_integers.norm` to map from any integral closure of `ℤ` in `K` to any integral closure of `ℤ` in `L`, in the same way that we don't prove things for `fraction_ring ℤ` but instead for any `is_fraction_ring`. Then we don't have these synonyms of `ℤ` lying around, which wreck all kinds of automation and typeclass synthesis.

riccardobrasca replied:
Ah yes, you're totally right!
